mmstudio/src/main/java/org/micromanager/data/internal/DefaultImage.java

```diff
@@ -144,81 +147,91 @@ public DefaultImage(TaggedImage tagged, Coords coords, Metadata metadata)
+    *
+    * Input pixels will be used directly (i.e., they are not copied).
+    *
+    * @param pixels Image pixels.  Should be a Java array of bytes or shorts (not null).
+    * @param format PropertyMap specifying image width, height, and pixelType (not null).
+    * @param coords Coords to be used for this new image (can be null).
+    * @param metadata Metadata to be used this new image (can be null).
+    * @throws IllegalArgumentException
+    */
    public DefaultImage(Object pixels, PropertyMap format, Coords coords,
          Metadata metadata) throws IllegalArgumentException {
-      Preconditions.checkNotNull(pixels);
-      Preconditions.checkNotNull(format);
-      metadata_ = metadata == null ? new DefaultMetadata.Builder().build() :
-            (DefaultMetadata) metadata;
-      coords_ = coords == null ? Coordinates.builder().build() : coords;
-      rawPixels_ = DirectBuffers.bufferFromArray(pixels);
-      pixelWidth_ = format.getInteger(PropertyKey.WIDTH.key(), 0);
-      pixelHeight_ = format.getInteger(PropertyKey.HEIGHT.key(), 0);
-      pixelType_ = format.getStringAsEnum(PropertyKey.PIXEL_TYPE.key(),
-            PixelType.class, null);
-      if (pixelWidth_ * pixelHeight_ * pixelType_.getBytesPerPixel() !=
-            rawPixels_.capacity() * pixelType_.getBytesPerComponent()) {
-         throw new IllegalArgumentException("Image width, height, and pixel type do not match pixel array size");
-      }
+      this(pixels,
+              format.getInteger(PropertyKey.WIDTH.key(), 0),
+              format.getInteger(PropertyKey.HEIGHT.key(), 0),
+              format.getStringAsEnum(PropertyKey.PIXEL_TYPE.key(),
+                      PixelType.class, null).getBytesPerPixel(),
+              format.getStringAsEnum(PropertyKey.PIXEL_TYPE.key(),
+                      PixelType.class, null).getNumberOfComponents(),
+              coords,
+              metadata);
    }
 
    /**
-    * @param pixels Assumed to be a Java array of either bytes or shorts.
+    * Generates a DefaultImage from pixels, minimal image info, and the
+    * supplied coords and metadata.
+    *
+    * Input pixels will be used directly (i.e., they are not copied).
+    *
+    * @param pixels Image pixels.  Should be a Java array of bytes or shorts (not null).
+    * @param coords Coords to be used for this new image (can be null).
+    * @param metadata Metadata to be used this new image (can be null).
+    * @throws IllegalArgumentException
     */
    public DefaultImage(Object pixels, int width, int height, int bytesPerPixel,
          int numComponents, Coords coords, Metadata metadata) 
          throws IllegalArgumentException {
-      metadata_ = (DefaultMetadata) metadata;
-      if (metadata_ == null) {
-         // Don't allow images with null metadata.
-         metadata_ = new DefaultMetadata.Builder().build();
-      }
-      coords_ = coords;
-
-      rawPixels_ = DirectBuffers.bufferFromArray(pixels);
```

Comment:
No, I was updating all other constructors in DefaultImage to stop using DirectBuffers.  This specific constructor is used by `DataManager.createImage(Object pixels, int width, int height, int bytesPerPixel, int numComponents, Coords coords, Metadata metadata);`. where the documentation says: "Generate a new Image with the provided pixel data, rules for interpreting that pixel data, coordinates, and metadata.", suggesting that pixel data are used without copying.  It is also used to insert blank images by displaywindow.imagej. 

I suggest deciding what the behavior of data().creatImage should be.  There definitely should be an internal method that re-uses pixels, and I would argue there also should an external one that does not copy.

Reply:
I'm not sure why "Generate a new Image" suggests no copying. The behavior of `createImage` has always been to copy, and changing this is not safe.

Anyway, how about `createImage` always making a copy and adding a `wrapImmutablePixels` that doesn't copy?

Reply:
"Generate a new Image with the provided pixel data" very much suggests that no copying takes place.  In general, the documentation has to be explicit about the desired behavior (again, I first had those functions copying pixels, but decided to change back to match the documentation).  Documentation to the convertTaggedImage functions in DataManager should also make clear whether or not pixel data are copied. 
I don't know what a `wrapImmutablePixels ` function should look like or what it should do. Why not add a flag to `createImage`, or if you do not like that make a `createImageCopy` that does the same as `createImage` but copies pixels?

Reply:
What I was suggesting was that `wrapImmutablePixels` should do what your current `createImage` implementation is doing, i.e., wrap the array without making a copy.

I'm proposing that it should be safe for users to use `Image` without necessarily understanding that it needs to be immutable (that would be a lot to expect from casual, first-time script writers). All of the methods were explicitly designed to protect such users (and other people that use code written by such users), which is why a copy was being made upon construction.

If the natural-sounding `createImage` has scary data integrity issues, I wouldn't trust any data handled by that system. I think most of the world would agree that scientific data integrity is important, and I'm sure you do, too, so I must be failing to convey how scary the possibilities are.

A flag has two downsides:
- Looking at the code, `createImage(foo, false)` conveys nothing until you look at the documentation, and I can't remember which of `true` or `false` skips the copy.
- Again, anything named `createImage` should not do dangerous things.

I'm proposing `wrapImmutablePixels` specifically because it sounds scary, which it what it is.